scale to zero doc (#356)
railwayapp/docs: src/docs/reference/scale-to-zero.md

---
title: Scale to Zero
---

> 

Reviewer: I'd add the Priority Boarding banner here, as it's a beta feature. E.g.:
<img width="726" alt="ss 2023-10-05 at 11 37 00 PM" src="https://github.com/railwayapp/docs/assets/24421625/6e8ba9e2-9700-40c9-92ef-d2b7dc9c59c7">

> Scale to Zero means reducing unnecessary costs by ensuring you only use what you need, when you need it.

Reviewer: I feel this sentence is way too broad and open to interpretation.

Suggestion - start with:
> App Sleeping allows you to increase the efficiency of resource utilization on Railway. It may reduce the resource usage cost of a [service](/develop/services) by ...

## What is App Sleeping?

App sleeping enables automated detection of an inactive [service](/develop/services) based on outbound traffic.

> Enabling App Sleep on a service tells Railway to stop the service when it is inactive, effectively reducing the overall cost to run it.

Reviewer: Suggestion: Replace the sentence above ("App sleeping enables automated ...") with the second one ("Enabling App Sleep on a service ..."), so you get a pyramid style of "what it does" -> "how it does it" (i.e. it tells railway to stop the service -> it does this by detecting traffic)

<Image src="https://res.cloudinary.com/railway/image/upload/v1696017787/events_gjkaob.png"
alt="App Sleeping in the Activity Feed"
layout="intrinsic"
width={700} height={460} quality={100} />

> ### Inactive Service Detection

Reviewer: Thoughts on making "inactive" / "active" into "slept" and "wake" for consistency? This feels like it may be semantically referring to the same thing using different terms

> Inactivity is based on outbound packets. If no packets are sent from the service for over 10 minutes or longer, the service is considered inactive.

Reviewer: Thoughts on clarifying what types of traffic this applies to?

It's both inbound/outbound and it may be confusing to only reference a single one

Author: It is truly only outbound traffic that can wake up a service.  We can clarify this if you think users will often get tripped up.  I'll leave it as-is for now, and we can figure out a good way to wordsmith it

@20k-ultra

> ### Reactivation

Reviewer: Suggestion: "Waking a Service Up"

If we call this "app sleeping" we should document the states it can be in, in a consistent manner (asleep or awake)

> When the service receives traffic from the _public network*_, it is sent a signal to wake.  

Reviewer: Also happens when a service sends traffic to WAN

> Only when the service is ready, is traffic routed to it, ensuring no requests are dropped during the startup process.
> 
> _*Currently, requests made to the service over the private network do not cause it to awake from sleep._

Reviewer: Suggestion:
```suggestion
We wake a service when it sends or receives traffic to/from the internet.

 The first request made to or sent from a service wakes a sleeping service up. This may take a small amount of time for the service to spin up again on the first request (commonly known as "cold boot time").
```

## How to enable App Sleeping

> To enable App Sleeping simply toggle the feature on within the service configuration - 

Reviewer: ```suggestion
To enable App Sleeping, toggle the feature on within the service configuration pane in your project:
```

<Image src="https://res.cloudinary.com/railway/image/upload/v1696548703/docs/scale-to-zero/appSleep_ksaewp.png"
alt="Enable App Sleep"
layout="intrinsic"
width={700} height={460} quality={100} />

> 1. Navigate to your service's settings > Deploy > App Sleeping.
> 2. Toggle "Enable App Sleeping".
> 3. This setting will be applied across all [replicas](https://docs.railway.app/develop/services#horizontal-scaling-with-replicas) automatically.
> 4. When your service is put to sleep, an event will be visible in the dashboard.
> 5. To _disable_ App Sleeping, simply toggle the setting again.

Reviewer: ```suggestion
1. Navigate to your service's Settings > Deploy > App Sleeping
2. Toggle "Enable App Sleeping"
3. To _disable_ App Sleeping, toggle the setting again
```

Reviewer: Reasoning/questions:
- "this setting will be applied to replicas" have nothing to do with enabling this feature; it's a caveat, not a step in enabling it
- "an event will be visible in the dashboard" - need to be more specific, but would also throw this _outside_ of the steps to enable (intermingling instructional & informational content may cause confusion)

> ## Things to Note

Reviewer: ```suggestion
## Caveats
```

Reviewer 2: This is funny because we did use Caveats but switched it to be more neutral wording instead of something that sounds negative. I'm ok with either route.

> - **Latency** - There may be a small delay when requests arrive at a stopped application.
> - **Private Networking** - Services that receive traffic exclusively over the private network should *not* be enabled with App Sleeping since only public network traffic will trigger the service to wake.
> - **Conditions for Sleeping** - For a service to be considered inactive, it must not send outbound traffic for at least 10 minutes or more.  Outbound traffic can include telemetry, NTP, etc.

Reviewer: ```suggestion
- There will be a small delay in the response time of the first request sent to a slept service (commonly known as "cold boot times")
- [Private Networking](/reference/private-networking) does not currently work with App Sleeping
- Outbound traffic is excluded from considering when to sleep a service. For Railway to put a service to sleep, a service must not send outbound traffic for at least 10 minutes. Outbound traffic can include telemetry, NTP, etc.
- Enabling App Sleeping will apply the setting across all [Replicas](https://docs.railway.app/develop/services#horizontal-scaling-with-replicas)
```
